mautic/mautic
# Add a new trigger for changing a lead stage (refactor stage setting)

| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | no
| New feature/enhancement? (use the a.x branch)      | yes
| Deprecations?                          | no
| BC breaks? (use the c.x branch)        | maybe
| Automated tests included?              | yes
| Related user documentation PR URL      |  N/A
| Related developer documentation PR URL | N/A
| Issue(s) addressed                     | N/A


#### Description:

This adds a new feature, where a user can change a contact's "Stage" when the contact reaches a certain level of points.

<img width="1029" alt="image" src="https://github.com/mautic/mautic/assets/13075514/a9a3bd79-4c60-4f5e-b161-f18d9e244375">

#### Steps to test this PR:

Testing new functionality:
1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
2. Set up a new Point > Action (so contacts can accumulate points)
3. Set up a new Point > Trigger with the Event "Change contacts Stage"
4. Set up a new Point > Trigger with an existing Event (e.g. add Tags)
5. Publish and apply the Trigger to the existing contacts: The new stage should be visible on all contacts that qualify based on their level of points. And the existing Event should also have fired (e.g. new tags).
6. Perform an action that allows the contact to reach the necessary level of points so that the Trigger can fire
7. When the contact reached the level, check if the level was changed as configured in step 3. And the existing Event should also have fired (e.g. new tags).

Testing existing functionality
1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester)).
2. Create a campaign with an action that changes the stage of a contact
3. Run the campaign and ensure the stage was changed as expected. 


#### Possible BC break:

IMO they are not BC breaks. I checked the Mautic code base. Tests are ok too. But I can not be a 100% sure.

**Renaming addToStages to addToStage**
I noticed that the name of the method [LeadModel::addToStages](https://github.com/mautic/mautic/blob/5.x/app/bundles/LeadBundle/Model/LeadModel.php#L998) is a bit misleading. The contact can only be added to one Stage. So I renamed it to `addToStage`. I fixed it everywhere in the code Mautic code, but I'm not sure about 3rd party plugins.

**Only allow a Lead object for LeadModel::addToStage()**
Before it was possible to pass in a Lead ID and the method would convert it to a Lead object. As far as I could see, this was not used in the code. So I wanted to make it simpler and removed that option. 
I added a new function `StageHelper::getStage($stageId)`
